pkg/prefetch/manager.go

```diff
@@ -81,7 +88,7 @@ func (m *CredentialManager) Run(ctx context.Context, parallelRoutines int) {
 func (m *CredentialManager) handleExpiring(ctx context.Context, credentials *sts.CachedCredentials) {
 	logger := log.WithFields(sts.CredentialsFields(credentials.Identity, credentials.Credentials))
 
-	active, err := m.IsRoleActive(credentials.Identity.Role)
+	active, err := m.IsRoleActive(credentials.Identity.Role.Name)
```

Comment:
I suspect this needs to change (and thus also needs some tests to prove the behaviour).

The purpose was originally to find any Pods in the Kubernetes API server filled cache. It should probably take the RoleIdentity type and then use that string for the cache key.

I'd expect the failing test to be something like if we had 2 pods using the same role but different session names, when one of them is no longer active (but the other is) we'd expect the non-active session to no longer be retained. If it's just matching on Role name, then it would keep them active.

The behaviour isn't horrific (it'd just mean refreshing credentials past them needing to be used) but be good to make it match the rest? What do you think?

Reply:
@pingles just so I have this correct, we want to change the pod cache to use something like `indexIdentity` over the existing `indexPodRole`? where the existing cache of pods by only the role changes to be pods cached by the identity?

Reply:
Yep, that's right. At the moment, I would expect us to have incorrect behaviour assuming a state of:

```
Pod{Name: "foo", Status: "active", Role: "reader", Session: "active-reader"}
Pod{Name: "bar", Status: "stopped", Role: "reader", Session: "stopped-reader"}
```

`IsRoleActive` for an identity of `Identity{Role: "reader", Session: "stopped-reader"}` would still return true (because the active Pod remains) despite the call checking a different identity.

I suspect we should also include `ExternalID` if that's included also?

Reply:
Ok got it, I am thinking of just changing to build the Identity and using `String` on that for the key, will pull out constructing an identity to an `IdentityFromPod` function and go from there. Hold my :beer:

Reply:
Sounds great to me 🤞

Reply:
Okies hopefully I did this right, refactoring+tests added :crossed_fingers: